core/internal/filetransfer/file_transfer_s3.go

```diff
@@ -0,0 +1,289 @@
+package filetransfer
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/aws-sdk-go-v2/service/s3/types"
+	"github.com/wandb/wandb/core/internal/observability"
+	"golang.org/x/sync/errgroup"
+)
+
+var maxS3Workers int = 500
+var s3Scheme string = "s3"
+
+// S3FileTransfer uploads or downloads files to/from s3
+type S3FileTransfer struct {
+	// client is the HTTP client for the file transfer
+	client *s3.Client
+
+	// logger is the logger for the file transfer
+	logger *observability.CoreLogger
+
+	// fileTransferStats is used to track upload/download progress
+	fileTransferStats FileTransferStats
+
+	ctx context.Context
```

Comment:
Each Task should have its own Context rather than there being a single long-lived Context; follow the example for `DefaultTask` which has a `Context` field (which isn't properly used at the moment).

Contexts are best used as "request"-scoped values: for web servers the request is obvious (it's the literal request that triggered a code path), and in our case a "request" should usually correspond to a message received from the user process (to be concrete: the `DownloadArtifactRequest`).

Although Contexts aren't really used at the moment, I'm working on a PR that will rely on FileTransfers using proper per-request Contexts.

Reply:
not having one long lived context makes sense. to clarify, each task should have its own context field, which is populated from the request made by the user process, right? I'll create a ticket for our team to pick this up